amaranth/ml/train.py

```diff
@@ -52,26 +57,70 @@ def main():
       low_calorie_threshold=amaranth.LOW_CALORIE_THRESHOLD,
       high_calorie_threshold=amaranth.HIGH_CALORIE_THRESHOLD)
 
+  # Normalize input strings
+  # Step 1: convert strings to lowercase
+  # Step 2: filter out characters present in DISH_NAME_FILTERS
+  # Step 3: re-combine characters back to a normal string
+  calorie_data['description'] = calorie_data['description'].apply(
+      lambda desc: desc.lower(),  # Step 1
+  ).apply(
+      lambda desc: [char for char in desc
+                    if char not in DISH_NAME_FILTERS],  # Step 2
+  ).apply(
+      ''.join,  #Step 3
+  )
+
   # Do some preprocessing and calculations for encoding
-  corpus = calorie_data['description'].str.replace(',', '').str.lower()
+  corpus = calorie_data['description']
   vocab_size = lib.num_unique_words(corpus)
   tokenized_corpus = corpus.map(lambda desc: desc.split(' '))
   max_corpus_length = lib.max_sequence_length(tokenized_corpus)
 
-  # Create TextVectorization layer
-  vectorize_layer = keras.layers.experimental.preprocessing.TextVectorization(
-      max_tokens=vocab_size,
-      output_sequence_length=max_corpus_length,
-  )
+  # Count appearances of each word in dataset
+  tokenizer_cnt = defaultdict(int)
+  for dish_name in calorie_data['description']:
+    for token in dish_name.split():
+      tokenizer_cnt[token] += 1
+
+  # Only 'remember' words that appear at least MIN_TOKEN_APPEARANCE times
+  keep_tokens = [
```

Comment:
nit: I know this makes things into one line, but it's honestly hard to read. Do you mind writing this as a for loop?

Reply:
I see how that could be hard to read, I'll refactor!

Reply:
Fixed in https://github.com/googleinterns/amaranth/pull/35/commits/783d9d36ecb4e226f0d1b8ee1d0bb1a187c37d10